### PR TITLE
[8.0] add purchase_add_product_supplierinfo

### PR DESCRIPTION
**Pull request by chafique-delli**
This module allows from the purchase order, to complete the supplierinfo list of each product.

### CODE REVIEW

File: purchase_add_product_supplierinfo/purchase.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+class PurchaseOrder(models.Model):
+    _inherit = 'purchase.order'
+
+    @api.model
+    def _check_product_supplierinfo(self):
+        lines = []
+        for line in self.order_line:
+            suppinfo = self.env['product.supplierinfo'].search([
+                ('product_id', '=', line.product_id.id),
+                ('name', '=', self.partner_id.id)])
```
Comment (sebastienbeau) [inline purchase_add_product_supplierinfo/purchase.py]: Wrong domain here.
If you have a generic supplier for all of the variant the product_id field of the supplier will be empty

Moreover if the self.partner_id.id is not the commercial entity you will never find the suplier

File: purchase_add_product_supplierinfo/purchase.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                    'product_id': line.product_id.id,
+                }))
+        return lines
+
+    @api.multi
+    def purchase_confirm(self):
+        self.ensure_one()
+        lines_for_update = self._check_product_supplierinfo()
+        if lines_for_update:
+            if self.partner_id.parent_id:
```
Comment (sebastienbeau) [inline purchase_add_product_supplierinfo/purchase.py]: I think there is a field or a method to get the commercial entity id. try self.partner_id._commercial_partner_id

**Comment by sebastienbeau** [comment]
@chafique-delli can you fis the way to get the right supplier and add a test with
- a specific supplier (a supplier info attach to one variant)
- a generic supplier (a supplier info attach to the template => all variant)

File: purchase_add_product_supplierinfo/purchase.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                default_wizard_line_ids=lines_for_update,
+            )
+            add_supplierinfo_form = self.env.ref(
+                'purchase_add_product_supplierinfo.'
+                'view_purchase_add_supplierinfo_form', False)
+            return {
+                'name': _("Associate the supplier '%s' with the products "
+                          "of this purchase order.") % supplier_id.name,
+                'type': 'ir.actions.act_window',
+                'view_type': 'form',
```
Comment (bealdav) [inline purchase_add_product_supplierinfo/purchase.py]: unseful key from 7.0

File: purchase_add_product_supplierinfo/README.rst [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+   :alt: License: AGPL-3
+
+============================================
+Add Product Supplierinfo from Purchase Order
+============================================
+
+This module allows in the purchase order, to automatically complete with
+partner, the supplierinfo list for all products which are not related
+to this supplier
+
```
Comment (bealdav) [inline purchase_add_product_supplierinfo/README.rst:12]: I do not understand the text above. 
What do you think of "complete missing entries (product, supplier) in supplier info list before purchase order confirmation" ?

**Comment by bealdav** [comment]
Thanks for this valuable module.

**Comment by gurneyalex** [comment]
@chafique-delli I fixed the runbot config and triggered a rebuild.

File: purchase_add_product_supplierinfo/tests/test_purchase_add_product_supplierinfo.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        # add product supplierinfo
+        wizard.with_context(active_id=purchase_8.id).add_product_supplierinfo()
+        product = self.env['product.product'].browse(
+            result8['context']['default_wizard_line_ids'][0][2]['product_id'])
+        supplierinfo_ids = self.env['product.supplierinfo'].search([
+            ('product_tmpl_id', '=', product.product_tmpl_id.id),
+            ('name', '=', purchase_8.partner_id.id)])
+        self.assertNotEquals(supplierinfo_ids, False)
+
+        # purchases with product supplierinfo to update and
```
Comment (sebastienbeau) [inline purchase_add_product_supplierinfo/tests/test_purchase_add_product_supplierinfo.py:43]: Can you split the test one test per purchase?

**Comment by chafique-delli** [comment]
@gurneyalex, thank you for the fix on the runbot config but there is another problem with runbot. Is it you can watch it please.

**Comment by gurneyalex** [comment]
@chafique-delli your module is somehow breaking the loading of demo data from this file: https://github.com/OCA/purchase-workflow/blob/8.0/purchase_requisition_auto_rfq/demo/product_and_supplier.yml

Could you check what is amiss?

**Comment by sebastienbeau** [comment]
Youhou ! it's green!
Great work Chafique, tested :+1:

**Comment by hparfr** [comment]
👍

File: purchase_add_product_supplierinfo/purchase.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        self.ensure_one()
+        lines_for_update = self._check_product_supplierinfo()
+        if lines_for_update:
+            if self.partner_id.commercial_partner_id:
+                supplier_id = self.partner_id.commercial_partner_id
+            else:
+                supplier_id = self.partner_id
+            ctx = dict(
+                default_wizard_line_ids=lines_for_update,
+            )
```
Comment (gurneyalex) [inline purchase_add_product_supplierinfo/purchase.py:40]: side note: this is not the efficient way of creating a dictionary in Python. It is better to use:

```
ctx = {'default_wizard_line_ids': lines_for_update}
```

Rationale: your version creates 2 dictionaries, one for the kwargs passed, and another one for the result, and the 1st one is then garbage collected, to it is better to use a litteral dictionary in this case.

**Review by gurneyalex — APPROVED** [review]

**Comment by gurneyalex** [comment]
merged in 645e5dfa5376ae49ead4cb9b07d6177852950516

**Pull request closed without merging**